### ISSUE TITLE
Done callback is raised before processed all lines.

### ISSUE THREAD

**Issue by sp9usb** [danielkrainas/node-netstat#4]
Done callback `options.done` is raised before processed all lines...

**Comment by mnkhouri** [comment]
I believe this is still an issue. I read through the fix that @danielkrainas pushed, but I believe it doesn't work around Node's behaviour. Here is a minimal test case that shows the problem:

```
# cat test.js 
var proc = require('child_process').spawn('netstat', ['-ane'])
proc.stdout.on('data', function () { console.log('data') })
proc.on('exit', function () { console.log('exit') })

# node test.js 
exit
data

# node --version
v0.10.42
```

The root of the issue is that the child process may emit `exit` _before_ the child process's stdout emits `data`.... I'm still trying to figure out a way to work around this, but my current plan is to switch the `spawn()` call to an `exec()` call.

Note that I'm using a really old Node version, on a platform with a very slow processor. This may not be an issue on newer versions / faster platforms.

**Comment by mnkhouri** [comment]
Running the same test on a much faster machine (but still using Node v0.10.42) does not show the same problem. Might not be worth it to try to find a fix. My personal fix is to do the following (use `exec()` instead of `spawn()`):

```
        this.netstat.activators.async = function (cmd, args, makeLineHandler, done) {
            var lineHandler = makeLineHandler(this.netstat.utils.noop)
            var fullCmd = cmd + ' ' + args.join(' ')
            exec(fullCmd, {timeout: 1000}, function(error, stdout, stderr) {
                if (error) {
                    done(error)
                }
                else {
                    var lines = this.netstat.utils.parseLines(stdout)
                    for (var i = 0; i < lines.length; i++) {
                        lineHandler(lines[i])
                    }
                    done(null)
                }
            }.bind(this))
        }.bind(this)
```

**Comment by danielkrainas** [comment]
The problem with using exec instead of spawn is in the case of long running netstat instances where you want to stream the output to your app to deal with properly as the netstat data is output, not all at once when it's completed. Do you find the same problem with `done()` being called prematurely if you replace the 'exit' event with 'close' instead? The close event happens when the stdio streams have been closed, but not always when the child process has exited.

**Comment by mnkhouri** [comment]
Yeah, you're right that exec has downsides.

Using the 'close' is a great idea -- it resolved the issue when I tested it both in my application and in the test case I posted above. Would you be willing to modify the library to use 'close' instead of 'exit'? If you'd like to keep using 'exit', I can also override the activator in my application (via PR #10 )

**Comment by danielkrainas** [comment]
Glad to hear that worked! I can make the change and publish a release. We don't really need to worry about 'exit' since node-netstat is only concerned with the stdio anyways. I'm going to close #10 in the meantime and reopen this to track.

**Comment by danielkrainas** [comment]
fixed by 8b03d859008d3bb487bf06762818592ef958d3f5 and available in [v1.3.1](https://github.com/danielkrainas/node-netstat/releases/tag/v1.3.1)
